learning_wavelets/models/exact_recon_unet.py

```diff
@@ -0,0 +1,169 @@
+                non_linearity=self.non_linearity,
+            ) for _ in range(self.n_non_lins)
+        ]
+
+    def call(self, inputs):
+        outputs = inputs
+        for conv in self.convs:
+            outputs = conv(outputs)
+        return outputs
+
+class UpConv(Layer):
+    def __init__(self, n_filters, kernel_size=3, **kwargs):
+        super().__init__(**kwargs)
+        self.n_filters = n_filters
+        self.kernel_size = kernel_size
+        self.conv = Conv2D(
+            filters=self.n_filters,
+            kernel_size=self.kernel_size,
+            padding='same',
+            activation=None,
+        )
+        self.up = UpSampling2D(size=(2, 2))
+
+    def call(self, inputs):
+        outputs = self.up(inputs)
+        outputs = self.conv(outputs)
+        return outputs
+
+
+class ExactReconUnet(Model):
+    def __init__(
+            self,
+            n_output_channels=1,
+            kernel_size=3,
+            layers_n_channels=[64, 128, 256, 512, 1024],
+            layers_n_non_lins=2,
+            non_linearity='relu',
+            **kwargs,
+        ):
+        super().__init__(**kwargs)
+        self.n_output_channels = n_output_channels
+        self.kernel_size = kernel_size
+        self.layers_n_channels = layers_n_channels
+        self.n_layers = len(self.layers_n_channels)
+        self.layers_n_non_lins = layers_n_non_lins
+        self.non_linearity = non_linearity
+        self.down_convs = [
+            ConvBlock(
+                n_filters=n_channels,
+                kernel_size=self.kernel_size,
+                non_linearity=self.non_linearity,
+                n_non_lins=self.layers_n_non_lins,
+            ) for n_channels in self.layers_n_channels[:-1]
+        ]
+        self.down = MaxPooling2D(pool_size=(2, 2), padding='same')
+        self.bottom_conv = ConvBlock(
+            n_filters=self.layers_n_channels[-1],
+            kernel_size=self.kernel_size,
+            non_linearity=self.non_linearity,
+            n_non_lins=self.layers_n_non_lins,
+        )
+        self.up_convs = [
+            ConvBlock(
+                n_filters=n_channels,
+                kernel_size=self.kernel_size,
+                non_linearity=self.non_linearity,
+                n_non_lins=self.layers_n_non_lins,
+            ) for n_channels in self.layers_n_channels[:-1]
+        ]
+        self.ups = [
+            UpConv(
+                n_filters=n_channels,
+                kernel_size=self.kernel_size,
+            ) for n_channels in self.layers_n_channels[:-1]
+        ]
+        self.final_conv = Conv2D(
+            filters=self.n_output_channels,
+            kernel_size=1,
+            padding='same',
+            activation=None,
+        )
+
+    
+    def call(self, inputs):
+        scales = []
+        noisy_image = inputs[0]
+        noise_std = inputs[1]
+        outputs = noisy_image
+        outputs, crop_w, crop_h, w, h = pad_power_of_two(outputs, self.layers_n_channels)
```

Comment:
this shouldn't depend on `self.layers_n_channels`.
This variable indicates how many channels there are per scale but not how many scales there are.

`n_layers` is the number of scales

Reply:
I see what you did afterwards, but I think it should directly use `self.n_layers`